cylc/flow/task_pool.py

```diff
@@ -299,14 +299,14 @@ def release_runahead_tasks(self):
 
         released = False
 
-        # Any finished tasks can be released immediately (this can happen at
-        # restart when all tasks are initially loaded into the runahead pool).
-        # And any manually-triggered task.
+        # At restart all tasks are loaded as runahead-limited, but finished and
+        # manually-triggered ones (including --start-task) can be released
+        # immediately. Note runahead release can cause the task pool to change
+        # size because we spawn parentless tasks on previous-instance release.
 
         for itask in (
             itask
-            for point_id_map in self.main_pool.values()
-            for itask in point_id_map.values()
+            for itask in self.get_tasks()
```

Comment:
This was actually a bug, because the task pool can change size during iteration (as described in the comment above).

Reply:
(In fact the above example triggers this bug, on master)

Reply:
I don't understand the difference with how it was previously, considering the definition of
https://github.com/cylc/cylc-flow/blob/ae6f0652e7ec656e20efc8940214ae30d3234eaf/cylc/flow/task_pool.py#L664-L672

(apart from the fact that `self.get_tasks()` can mutate `self.main_pool_list` and `self.main_pool_changed`)

Reply:
If I run the example above, with `--start-task=a.2`:
```
2021-06-23T10:17:47+12:00 INFO - [a.2] -released from runahead
2021-06-23T10:17:47+12:00 INFO - (a.2) spawned a.3 flow(i)
2021-06-23T10:17:47+12:00 ERROR - dictionary changed size during iteration
        Traceback (most recent call last):
          File "/home/oliverh/cylc/cylc-flow/cylc/flow/scheduler.py", line 572, in start_scheduler
            await self.main_loop()
          File "/home/oliverh/cylc/cylc-flow/cylc/flow/scheduler.py", line 1409, in main_loop
            if not self.is_paused and self.pool.release_runahead_tasks():
          File "/home/oliverh/cylc/cylc-flow/cylc/flow/task_pool.py", line 306, in release_runahead_tasks
            for itask in (
          File "/home/oliverh/cylc/cylc-flow/cylc/flow/task_pool.py", line 306, in <genexpr>
            for itask in (
        RuntimeError: dictionary changed size during iteration
2021-06-23T10:17:47+12:00 CRITICAL - Workflow shutting down - dictionary changed size during iteration
```

Reply:
On this branch, we iterate over the list of tasks (returned by `self.get_tasks()`) that reflects the state of the pool before iteration begins.

On master, we iterate over `main_pool.values()`. In Python 3, `dict.values()` does not return a list of values, it's a "view" of the actual dictionary, so modifying it in-flight is bad.  https://docs.python.org/3/library/stdtypes.html#dict-views


```
for point_id_map in self.main_pool.values():                                                                                                                                                                                                   
    for itask in point_id_map.values():   
        #...
```

Reply:
I found this bug too:
```
2021-06-30T12:27:08+12:00 INFO - [a.2] -released from runahead
2021-06-30T12:27:08+12:00 INFO - (a.2) spawned a.3 flow(r)
2021-06-30T12:27:08+12:00 ERROR - dictionary changed size during iteration
        Traceback (most recent call last):
          File "/home/sutherlander/cylc/cylc-flow/cylc/flow/scheduler.py", line 572, in start_scheduler
            await self.main_loop()
          File "/home/sutherlander/cylc/cylc-flow/cylc/flow/scheduler.py", line 1405, in main_loop
            if not self.is_paused and self.pool.release_runahead_tasks():
          File "/home/sutherlander/cylc/cylc-flow/cylc/flow/task_pool.py", line 309, in release_runahead_tasks
            for point_id_map in self.main_pool.values()
          File "/home/sutherlander/cylc/cylc-flow/cylc/flow/task_pool.py", line 308, in <genexpr>
            itask
        RuntimeError: dictionary changed size during iteration
2021-06-30T12:27:08+12:00 CRITICAL - Workflow shutting down - dictionary changed size during iteration
```
Changed it to:
```
         for itask in (
             itask
-            for point_id_map in self.main_pool.values()
+            for point_id_map in list(self.main_pool.values())
             for itask in point_id_map.values()
             if itask.state.is_runahead
             if itask.state(

```
For the purposes of reviewing this:
```
2021-06-30T12:33:26+12:00 INFO - (a.2) spawned b.2 flow(w)
2021-06-30T12:33:26+12:00 INFO - [a.3] status=running: (received)succeeded at 2021-06-30T12:33:25+12:00  for job(01) flow(w)
2021-06-30T12:33:26+12:00 INFO - (a.3) spawned b.3 flow(w)
2021-06-30T12:33:26+12:00 INFO - Workflow shutting down - AUTOMATIC
2021-06-30T12:33:26+12:00 WARNING - Partially satisfied prerequisites left over:
        b.2 is waiting on:
        * b.1 succeeded
ESC[0m2021-06-30T12:33:26+12:00 INFO - DONE
```

Reply:
Ah, I'm slightly confused @dwsutherland - isn't that exactly the same bug that I fixed *on this branch*?? (although you've fixed it in a slightly different way, same result).

Reply:
My bad should have phrased it:

"While trying to reproduce the problem on master I found that same bug too:"

Yes, it's fixed on this branch.. But it wasn't the purpose of this branch (though clearly related), so when trying to reproduce the master branch I ran into the same bug that is fixed in this branch.....

It was important to fix master to test the actual purpose of this PR, to know how master handles pre-initial dependencies (because we can't even get that far in the presence of the bug master)... So I made that above temporary fix.

Reply:
OK, no worries - thanks for clarifying!